bio2zarr/ts.py

```diff
@@ -0,0 +1,271 @@
+        self.ts = tskit.load(ts_path)
+        self.contig_id = contig_id if contig_id is not None else "1"
+        self.isolated_as_missing = isolated_as_missing
+        self.root_attrs = {}
+
+        self._make_sample_mapping(ploidy)
+        self.contigs = [vcz.Contig(id=self.contig_id)]
+        self.num_records = self.ts.num_sites
+        self.positions = self.ts.sites_position
+
+    def _make_sample_mapping(self, ploidy):
+        ts = self.ts
+        self.individual_ploidies = []
+        self.max_ploidy = 0
+
+        if ts.num_individuals > 0 and ploidy is not None:
+            raise ValueError(
+                "Cannot specify ploidy when individuals are present in tables"
+            )
+
+        # Find all sample nodes that reference individuals
+        individuals = np.unique(ts.tables.nodes.individual[ts.samples()])
+        if len(individuals) == 1 and individuals[0] == tskit.NULL:
+            # No samples refer to individuals
+            individuals = None
+        else:
+            # np.unique sorts the argument, so if NULL (-1) is present it
+            # will be the first value.
+            if individuals[0] == tskit.NULL:
+                raise ValueError(
+                    "Sample nodes must either all be associated with individuals "
+                    "or not associated with any individuals"
+                )
+
+        if individuals is not None:
+            self.sample_ids = []
+            for i in individuals:
+                if i < 0 or i >= self.ts.num_individuals:
+                    raise ValueError("Invalid individual IDs provided.")
+                ind = self.ts.individual(i)
+                if len(ind.nodes) == 0:
+                    raise ValueError(f"Individual {i} not associated with a node")
+                is_sample = {ts.node(u).is_sample() for u in ind.nodes}
+                if len(is_sample) != 1:
+                    raise ValueError(
+                        f"Individual {ind.id} has nodes that are sample and "
+                        "non-samples"
+                    )
+                self.sample_ids.extend(ind.nodes)
+                self.individual_ploidies.append(len(ind.nodes))
+                self.max_ploidy = max(self.max_ploidy, len(ind.nodes))
+        else:
+            if ploidy is None:
+                ploidy = 1
+            if ploidy < 1:
+                raise ValueError("Ploidy must be >= 1")
+            if ts.num_samples % ploidy != 0:
+                raise ValueError("Sample size must be divisible by ploidy")
+            self.individual_ploidies = np.full(
+                ts.num_samples // ploidy, ploidy, dtype=np.int32
+            )
+            self.max_ploidy = ploidy
+            self.sample_ids = np.arange(ts.num_samples, dtype=np.int32)
+
+        self.num_samples = len(self.individual_ploidies)
+
+        self.samples = [vcz.Sample(id=f"tsk_{j}") for j in range(self.num_samples)]
+
+    def iter_alleles(self, start, stop, num_alleles):
```

Comment:
Ah yes, it's a bit unfortunate we have to do this. I can see ways to work around though.